### PR TITLE
octopus: rbd-mirror: synchronize with in-flight stop in ImageReplayer::stop()

### PR DESCRIPTION
**Pull request by pponnuvel**
backport tracker: https://tracker.ceph.com/issues/54377

---

backport of https://github.com/ceph/ceph/pull/45106 and https://github.com/ceph/ceph/pull/45897
parent tracker: https://tracker.ceph.com/issues/54344

this backport was staged using ceph-backport.sh version 16.0.0.6848
find the latest version at https://github.com/ceph/ceph/blob/master/src/script/ceph-backport.sh

### CODE REVIEW

**Comment by ideepika** [comment]
jenkins retest this please

**Comment by sunnyku** [comment]
jenkins retest this please

**Comment by pponnuvel** [comment]
jenkins retest this please

**Comment by pponnuvel** [comment]
jenkins test make check

**Comment by idryomov** [comment]
This is one of the new test cases:
```
[ RUN      ] TestMockImageReplayer.StopJoinInterruptedReplayer
/home/jenkins-build/build/workspace/ceph-pull-requests/src/common/Timer.cc: In function 'Context* SafeTimer::add_event_after(double, Context*)' thread 7fb086117700 time 2022-02-28T11:05:36.994225+0000
/home/jenkins-build/build/workspace/ceph-pull-requests/src/common/Timer.cc: 117: FAILED ceph_assert(((lock).is_locked()))
 ceph version Development (no_version) octopus (stable)
 1: (ceph::__ceph_assert_fail(char const*, char const*, int, char const*)+0x1a1) [0x7fb0a17f91f5]
 2: (()+0x159c493) [0x7fb0a17f9493]
 3: (SafeTimer::add_event_after(double, Context*)+0x48) [0x7fb0a179e798]
 4: (()+0x179d859) [0x55e3bfca2859]
 5: (()+0x17a935e) [0x55e3bfcae35e]
 6: (std::function<void (Context*)>::operator()(Context*) const+0x4d) [0x55e3bfc7f9e5]
 7: (decltype (((forward<std::function<void (Context*)> const&>)({parm#1}))((get<0ul>)((forward<std::tuple<Context*> >)({parm#2})))) testing::internal::ApplyImpl<std::function<void (Context*)> const&, std::tuple<Context*>, 0ul>(std::function<void (Context*)> const&, std::tuple<Context*>&&, testing::internal::IndexSequence<0ul>)+0x46) [0x55e3bfc7b55d]
 8: (decltype (ApplyImpl((forward<std::function<void (Context*)> const&>)({parm#1}), (forward<std::tuple<Context*> >)({parm#2}), (testing::internal::MakeIndexSequence<std::tuple_size<std::tuple<Context*> >::value>)())) testing::internal::Apply<std::function<void (Context*)> const&, std::tuple<Context*> >(std::function<void (Context*)> const&, std::tuple<Context*>&&)+0x4a) [0x55e3bfc7565f]
 9: (testing::Action<void (Context*)>::Perform(std::tuple<Context*>) const+0x53) [0x55e3bfc6ed4d]
 10: (testing::internal::ActionResultHolder<void>* testing::internal::ActionResultHolder<void>::PerformAction<void (Context*)>(testing::Action<void (Context*)> const&, testing::internal::Function<void (Context*)>::ArgumentTuple&&)+0x55) [0x55e3bfd2b80c]
 11: (testing::internal::FunctionMocker<void (Context*)>::UntypedPerformAction(void const*, void*) const+0x61) [0x55e3bfd1abb7]
 12: (testing::internal::UntypedFunctionMockerBase::UntypedInvokeWith(void*)+0x456) [0x55e3c0287dba]
 13: (testing::internal::FunctionMocker<void (Context*)>::Invoke(Context*)+0x54) [0x55e3bfcdd2a4]
 14: (rbd::mirror::image_replayer::MockReplayer::shut_down(Context*)+0x5b) [0x55e3bfcd012d]
 15: (()+0x17a8cab) [0x55e3bfcadcab]
 16: (()+0x17b5858) [0x55e3bfcba858]
 17: (Context::complete(int)+0x2b) [0x55e3bfc592d1]
 18: (ContextWQ::process(Context*)+0xe1) [0x55e3c0453863]
 19: (ThreadPool::PointerWQ<Context>::_void_process(void*, ThreadPool::TPHandle&)+0x36) [0x55e3c04555e8]
 20: (ThreadPool::worker(ThreadPool::WorkThread*)+0x731) [0x7fb0a17b5311]
 21: (ThreadPool::WorkThread::entry()+0x36) [0x7fb0a17b93e6]
 22: (Thread::entry_wrapper()+0x7c) [0x7fb0a17911d0]
 23: (Thread::_entry_func(void*)+0x1c) [0x7fb0a179114a]
 24: (()+0x9609) [0x7fb0aabf5609]
 25: (clone()+0x43) [0x7fb09fe0d293]
```

**Review by idryomov — COMMENTED** [review]
The backport itself seems fine but one of the new test cases fails.  This needs to be investigated, we may be missing something on the mocking support side.

**Comment by pponnuvel** [comment]
> The backport itself seems fine but one of the new test cases fails. This needs to be investigated, we may be missing something on the mocking support side.

I am not familiar with the mock tests. 
The files in `src/test/rbd_mirror/mock` directory are identical in master and this branch; so does the source file `test_mock_ImageReplayer.cc`. Is there anything I could check?

**Comment by idryomov** [comment]
Diffing `rbd_mirror`-related files is unlikely to help since the failure is in `common/SafeTimer.cc`.  If it was me, I'd take a hard look at the assert, the stack trace and the surrounding code.

**Comment by MrFreezeex** [comment]
I had the same issue on my dev box (on stream 8 with gcc 8.5.0) on https://github.com/ceph/ceph/pull/43359 and tested on a fresh master and I also had the same issue (even retrying a few time it doesn't pass)... So maybe there is a nasty bug that change the behavior with different gcc/lib versions?

**Comment by idryomov** [comment]
Did you try it on another machine or with another compiler by any chance?  I haven't had the time to look into it yet, just trying to collect some evidence...

**Comment by ideepika** [comment]
> Did you try it on another machine or with another compiler by any chance? I haven't had the time to look into it yet, just trying to collect some evidence...

@idryomov I am hitting this issue on vossi03 as well, will take a look.

**Comment by idryomov** [comment]
So `CentOS Stream 8` with `g++ (GCC) 8.5.0 20210514 (Red Hat 8.5.0-10)`, same as @MrFreezeex?

**Comment by idryomov** [comment]
Appended a fix for the unit test failures.

**Comment by idryomov** [comment]
jenkins test this please

**Comment by idryomov** [comment]
jenkins test this please

**Comment by idryomov** [comment]
jenkins test docs

**Comment by pponnuvel** [comment]
jenkins test docs

**Comment by pponnuvel** [comment]
jenkins retest this please

**Comment by pponnuvel** [comment]
jenkins test docs

**Comment by idryomov** [comment]
@pponnuvel The jinja2 issue probably affects other PRs as well, the fix should be separated into its own PR.

Also, I don't see a link to the corresponding master PR -- is master and other release branches not affected or has this issue been fixed in some other way there?

**Comment by pponnuvel** [comment]
@ideepika 
There's a PR https://github.com/ceph/ceph/pull/45839 already open for fixing dependencies including Jinja2. 

I am just testing if that resolves all issues here. Once confirmed, I'll then revert those changes.

So https://github.com/ceph/ceph/pull/45839 needs to be merged first. I think it affects all the Octopus PRs.

**Comment by pponnuvel** [comment]
jenkins render docs

**Comment by pponnuvel** [comment]
jenkins test docs

**Comment by pponnuvel** [comment]
jenkins retest this please

**Comment by pponnuvel** [comment]
Reverted the dependencies commits.

Checks are going to fail of course (until #45839 and #45972 are resolved).

**Comment by ideepika** [comment]
jenkins test docs please

**Comment by ljflores** [comment]
jenkins test docs

**Comment by pponnuvel** [comment]
@ljflores I think https://github.com/ceph/ceph/pull/45839 is needed as well.

jinja2 error:
```
ImportError: cannot import name 'environmentfilter' from 'jinja2' (/home/jenkins-build/build/workspace/ceph-pr-docs/build-doc/virtualenv/lib/python3.8/site-packages/jinja2/init.py)
```

`readthedocs-sphinx-search` no longer has branch named `master` (renamed to `main`):
```
Collecting git+https://github.com/readthedocs/readthedocs-sphinx-search@master (from -r admin/doc-read-the-docs.txt (line 2))
  Cloning https://github.com/readthedocs/readthedocs-sphinx-search (to revision master) to /tmp/pip-req-build-i1wgihsi
  Running command git clone --filter=blob:none --quiet https://github.com/readthedocs/readthedocs-sphinx-search /tmp/pip-req-build-i1wgihsi
  WARNING: Did not find branch or tag 'master', assuming revision or ref.
  Running command git checkout -q master
  error: pathspec 'master' did not match any file(s) known to git.
  error: subprocess-exited-with-error
  
  × git checkout -q master did not run successfully.
  │ exit code: 1
  ╰─> See above for output.
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
error: subprocess-exited-with-error

× git checkout -q master did not run successfully.
│ exit code: 1
╰─> See above for output.

note: This error originates from a subprocess, and is likely not a problem with pip.
Command time: 1s Return: 1
```

**Comment by ljflores** [comment]
@pponnuvel try rebasing the branch and force-pushing to re-trigger the `ceph-pr-docs` test (unless there's a better way to do this that I'm unaware of). The "readthedocs" failure is being addressed in both #46219 and #45839, so we'll have to wait until those are resolved.

**Comment by pponnuvel** [comment]
jenkins test make check

**Comment by pponnuvel** [comment]
jenkins test make check

**Comment by pponnuvel** [comment]
Thanks, @ljflores. Now just the "readthedocs" failure is left.

**Comment by ljflores** [comment]
jenkins render docs

**Comment by ljflores** [comment]
Hmm, `jenkins render docs` doesn't re-trigger that job. @pponnuvel rebase one more time because #46219 was merged just a few hours ago. That should fix "readthedocs".

**Comment by pponnuvel** [comment]
Yep, all fine now. @ljflores

**Comment by yuriw** [comment]
this passed tests and can be merged at will after approval
@pponnuvel @idryomov @ideepika 
https://trello.com/c/u7H7zc8R

**Review by idryomov — APPROVED** [review]

**Pull request merged**
